docxcompose/properties.py

```diff
@@ -198,96 +199,197 @@ def set_properties(self, properties):
+
+    def _get_fieldname_string(self):
+        raise NotImplementedError()
+
+    def _get_fieldname(self):
+        match = self.fieldname_search_expr.search(self._get_fieldname_string())
+        if match is None:
+            return None
+        return match.groups()[0]
+
+
+class RunNode(object):
+
+    def __init__(self, node, is_begin=False, is_end=False,
+                 is_separate=False, is_after_separate=False):
+        self.node = node
+        self.is_begin = is_begin
+        self.is_end = is_end
+        self.is_separate = is_separate
+        self.is_after_separate = is_after_separate
+
+
+class SimpleField(FieldBase):
+
+    attr_name = "{{{}}}instr".format(NS["w"])
+
+    def _get_fieldname_string(self):
+        return self.node.attrib[self.attr_name]
+
+    def update(self, value):
+        text = xpath(self.node, './/w:t')
+        if text:
+            text[0].text = self._format_value(value)
+
+    def remove_field_keep_value(self):
+        parent = self.node.getparent()
+        index = list(parent).index(self.node)
+        w_r = deepcopy(self.node[0])
+        parent.remove(self.node)
+        parent.insert(index, w_r)
+
+
+class ComplexField(FieldBase):
+
+    def __init__(self, field_node):
+        super(ComplexField, self).__init__(field_node)
+        # run and paragraph containing the field
+        self.w_r = self.node.getparent()
+        self.w_p = self.w_r.getparent()
+
+    def _get_fieldname_string(self):
+        return self.node.text
+
+    def _get_field_bound_indexes(self):
+        """Searches for the begin and end tags surrounding the complex
+        field as well as the corresponding position of the separate tag.
+        Returns their position in the parent node"""
+        fieldname_index = self.w_p.index(self.w_r)
+        begins = [self.w_p.index(el.getparent()) for el in xpath(self.w_p, u'.//w:r/w:fldChar[@w:fldCharType="begin"]')]
+        separates = [self.w_p.index(el.getparent()) for el in xpath(self.w_p, u'.//w:r/w:fldChar[@w:fldCharType="separate"]')]
+        ends = [self.w_p.index(el.getparent()) for el in xpath(self.w_p, u'.//w:r/w:fldChar[@w:fldCharType="end"]')]
+        for begin, separate, end in zip(begins, separates, ends):
+            if begin < fieldname_index < end:
+                return begin, separate, end
+        return None
+
+    def get_runs(self):
+        """Get run fields from the current complex field"""
+        runs = []
+        bounds = self._get_field_bound_indexes()
+        if bounds is None:
+            return runs
+        begin, separate, end = bounds
+
+        for run in xpath(self.w_p, u'.//w:r'):
+            position = self.w_p.index(run)
+            if begin <= position <= end:
+                runs.append(RunNode(run, position == begin, position == end,
+                                    position == separate, position > separate))
+        return runs
+
+    def update(self, value):
+        runs = self.get_runs()
```

Comment:
So what we are actually interested in here are only the `following-siblings` of the run containing the docproperty that come after the closest `separate`-run and before the closest `end`-run. I wonder, would it be simpler to only return a reference to a list of those runs? Then we'd know if the list contains only one run, we can just update that run, if there are more we can update the first one and drop the remaining runs.